Support GotoDefinition in LINQ query syntax

MaStr11 opened the pull request:
**Customer scenario**

Support *GotoDefinition* for LINQ query syntax. Enables Ctrl+Click and tool-tip info for LINQ query clauses. Proposal #23394.

*GotoDefinition* and *QuickInfo* are finding the right overloads in `System.Linq.Enumerable`,  `System.Linq.Queryable` and the  [query expression pattern](https://github.com/ljw1004/csharpspec/blob/gh-pages/expressions.md#the-query-expression-pattern) depending on the source.

![gotodefinitionlinq](https://user-images.githubusercontent.com/24472128/33327545-0d2bb59c-d458-11e7-9aaf-0ad7e030cf5f.gif)

(Use [gif-explode](http://gif-explode.com/?explode=https://user-images.githubusercontent.com/24472128/33327545-0d2bb59c-d458-11e7-9aaf-0ad7e030cf5f.gif) to view the single frames).

**Bugs this fixes:**

Feature request #22096 by @dpoeschl.

**Workarounds, if any**

New feature.

**Risk**

Performance degradation in the editor. VS crashes.

**Performance impact**

Unknown.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

Support for LINQ query syntax was already included but was missing some features (only `select` and `group by` was supported and `UnoptimizedForm` of the query was not taken into account as fall-back).

**How was the bug found?**

Feature request #22096 and proposal #23394.

**Test documentation updated?**

No.

MaStr11 commented:
Here is the test code from above for copy&paste including an `IQueryable` sample:
```CSharp
            var list = new Tuple<int, int>[][] {
                       new Tuple<int, int>[] { new Tuple<int, int>(1, 1), new Tuple<int, int>(2, 2) },
                       new Tuple<int, int>[] { new Tuple<int, int>(1, 2), new Tuple<int, int>(2, 1) } };
            var numbers = new int[] { 1, 2, 3, 4 };

            var q1 = from l1 in list
                     from l2 in l1
                     let even = (l2.Item1 % 2) == 0
                     where even
                     orderby l2
                     select new { number = l2, isEven = even };
            var q2 = from l1 in list
                     from l2 in l1
                     orderby l2.Item1, l2.Item2 descending
                     select l2;
            var q3 = from l1 in list
                     from l2 in l1
                     group l2 by l2.Item1;
            var q4 = from l1 in list
                     from l2 in l1
                     group l2 by l2.Item1 into g
                     select new { g.Key, Sum = g.Sum(t => t.Item2) };
            var q5 = from l1 in list
                     from l2 in l1
                     join n in numbers on l2.Item1 equals n
                     select n;
            var q6 = from l1 in list
                     from l2 in l1
                     join n in numbers on l2.Item1 equals n into g
                     select new { l1, g };

            var dc = new System.Data.Linq.DataContext("");
            var iqueryable = dc.GetTable<object>();
            var q7 = from q in iqueryable
                     select q;
```

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs:
```diff
@@ -3579,7 +3579,7 @@ void Goo()
         {
             await TestInMethodAsync("var q = from e in \"\" let $$y = 1 let a = new { y } select a;",
-                MainDescription($"({FeaturesResources.range_variable}) int y"));
+                MainDescription($"({FeaturesResources.range_variable}) int y (+ 1 {FeaturesResources.overload})"));
```
this seems to be a regression.

CyrusNajmabadi commented:
> Before: SymbolInfo for Select was null in cases where the select clause is not translated into a call to the Select method due to optimizations.
Now: SymbolInfo for the Select method is returned from BoundQueryClause.UnoptimizedForm

"for this Select method"?  Which "Select" method would this be?

CyrusNajmabadi commented:
> Tests are missing. Should these go into GoToDefinitionTests.vb or are there other places? 

Yes, they should go in:

Open\src\EditorFeatures\Test2\GoToDefinition\GoToDefinitionTests.vb

CyrusNajmabadi commented:
> Support for Visual Basic?

Yes.  Definitely necessary.

MaStr11 commented:
@CyrusNajmabadi  thanks for the review. I would like us to clarify some general questions before we proceed.

1. Is showing *quick info* for LINQ beneficial for the majority of C# programmers?

I (accidentally) enabled quick info for LINQ but I wonder whether this is really helpful (let, group and join translation are quite confusing). I also couldn't find any request for such a feature. The only thing I could find was [this opposing comment](https://github.com/dotnet/roslyn/issues/3529#issuecomment-112518021) by @DustinCampbell.
If we really want to do quick info for LINQ we need to define the desired end user behavior for this four portions of the [spec](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#query-expressions):

2. Degenerate query expressions
 
>A later phase of the translation removes degenerate queries introduced by other translation steps by replacing them with their source.

Example:

 ```CSharp
from c in customers.Where(c => c.City == "London")
select c

customers.Where(c => c.City == "London")
```

I see three options for quick info on the 'select' clause here:
* Don't show quick info.
* Show quick info just with a text that tells that the 'select' clause isn't mapped to a method due to optimizations.
* Show quick info for the unoptimized form (The `Select` method). (Info is taken from the unoptimized form calculated in [`Binder.BindQueryInternal2`](http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis.CSharp/Binder/Binder_Query.cs,105)). Current behavior of the PR.
* Show quick info for the method that is causing the optimization (in this case 'Where').

VS currently shows the type info for IEnumerable&lt;T&gt; which I think is unfortunate.

3. `select` clause preceded by a `join` clause

>A query expression with a join clause without an into followed by a select clause

```CSharp
from x1 in e1
join x2 in e2 on k1 equals k2
select v
```

>is translated into

```CSharp
( e1 ) . Join( e2 , x1 => k1 , x2 => k2 , ( x1 , x2 ) => v )
```

(There other clauses equivalent to this, e.g. `GroupJoin`)

I see three options for the 'select' clause here:
* Don't show quick info.
* Show quick info just with a text that tells that the 'select' clause isn't mapped to a method.
* Show quick info for the method that is consuming the select clause (in this case 'Join').

VS and this PR currently show the type info for IEnumerable&lt;T&gt; which I think is unfortunate.

4. `orderby` clause

>A query expression with an orderby clause

```CSharp
from x in e
orderby k1 , k2 , ..., kn
...
```

>is translated into

```CSharp
from x in ( e ) . 
OrderBy ( x => k1 ) . 
ThenBy ( x => k2 ) .
... .
ThenBy ( x => kn )
...
```

How should the method chain be shown in quick info? Should quick info above a comma show the method call?

sharwell commented:
@kuhlenh Can you get the design questions from @MaStr11 added to our meeting?

sharwell commented:
@MaStr11 During the design meeting, no one had strong opinions about what the behavior should be in any of the proposed cases. We would like to see a proposal written (by you) with a feature specification describing the behavior of the editor in normal cases as well as the cases written above. The most likely outcome of such a proposal is everyone would agree that it is fine and you can proceed with implementing it as described.

In other words:

1. We care more about making sure that the feature specification and implementation are consistent with each other than what the behavior actually is
2. There is no internal pressure to create this feature, so normally it wouldn't even come up; but if you are already interested in working on it then we don't have objections to that

MaStr11 commented:
Hi @sharwell,

This is my first attempt for a proposal. Could you please tell me what you think about it, what is missing or needs clarification? How will we proceed with the proposal? Shall I edit it if there are remarks (loosing history) or shall I post new versions or should we turn it into a gist or something to keep the history? 

As a general guidance for the proposal I tried to stay close to the [query expressions section of the C# spec](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#query-expressions). That means quick info should show the actual method mapping after all the transformations are completed.

Some general rules:
* quick info for parameter info is kept as is (the current PR has some regressions here that need to be addressed)
* quick info for clauses that directly map to methods is enhanced by this PR (The shipping version of VS already supports quick info for some query syntax clauses)
* for those clauses that don't directly map to a method a hint is shown: "This clause isn't mapped to an method call due to optimizations." or if this is not possible no quick info is shown.
* Goto definition is only supported for clauses that directly map to a method (ironically this will left the issue #22096 that started all this open as there are circumstances that prevent goto definition from working on some select clauses)

These clauses map directly to method calls (**bold** are the clauses that show quick info).

| ID | Topic      | Example                                                                  | Quick info               | Goto definition | 
|----|------------|--------------------------------------------------------------------------|--------------------------|-----------------| 
| 1  | SelectMany | from l1 in l<br> **from** l2 **in** l1                                   | Select method            | yes             | 
| 2  | Where      | from l1 in l<br> **where** true                                          | Where method             | yes             | 
| 3  | Group by   | from l1 in l<br> **group** l1 **by** l.Item1 into g                      | Group method             | yes             | 
| 4  | Join       | from l1 in l<br> **join** n **in** numbers **on** l1 **equals** n into g | Join or GroupJoin method | Yes             | 
| 5  | Select (only if mapped to a Select method)    | from l1 in l<br>**select** l1         | Select method (see also below) | Yes       |
| 6  | Select for let | from l1 in l<br>**let** v=true<br>select v                           | Select method (see also below) | Yes       |
| 7  | From variable types | **from** double d **in** l1                           | Cast method              | Yes             |
| 8  | Join variable types | **join** T x **in** e **o**n k1 **equals** k2            | :exclamation: Cast method is shadowed by call to GroupJoin. Quick info for GroupJoin is shown.             | No, opens GroupJoin instead            |

### Degenerated methods

e.g.

from i in l
where true
**select** i

Quick info for `select` shows "This clause isn't mapped to an method call due to optimizations." or nothing at all. Goto definition is not available.

### select clause preceded by a join clause and other similar clauses 

e.g.

from x1 in e1
join x2 in e2 on k1 equals k2
**select** v

Quick info for `select` shows "This clause isn't mapped to an method call due to optimizations." or nothing at all. Goto definition is not available.

### orderby clause

Quick info and goto definition are available for parameters with an corresponding `orderby`, `ascending` or `descending` clause. That means that the `ThenBy` mappings are only accessible via quick info if the `ascending` or `descending` clauses are present. The orderby clause is a fall-back in the case the first parameter does not have an `ascending` or `descending` clause:

| Clause                                   | This method call mapping is shown                                        |
|------------------------------------------|--------------------------------------------------------------------------| 
| **orderby** a                            | mapping for parameter a                                                  | 
| orderby a **ascending**                  | mapping for parameter a                                                  | 
| **orderby** a, b                         | mapping for parameter a, ThenBy mapping for b is not accessible via quick info | 
| **orderby** a, b **ascending**           | mapping for parameter a and b                                            | 
| orderby a **ascending**, b **ascending** | mapping for parameter a and b                                            | 

### let clause

Quick info reveals some transformation internals here (See also *Transparent identifiers* below). On the one hand side this is a bad thing because it unveils some intermediate types that are not really useful for the end user. But on the other side it illustrates that the let clause is nothing but a fancy select. (I do remember searching for a `Let` extension method on `Enumerable` when I used `let` for the first time.) Quick info would lighten up what is going on under the hood.

### Explicit range variable types (see cases 7 and 8)

Explicit range variable are translated into a call to `Cast`. 
❓ Only the `from` clause is supported because in the `join` clause the `GroupJoin` call overshadows the `Cast` call (There is no place where the Cast call could be shown).

### Transparent identifiers

Transparent identifiers are leaking through. While the names of the transparent identifiers are not revealed they leave traces as there are anonymous types visible in quick info and so the end user gets a glimpse at the magic that is going on. In the (slightly altered) sample from the spec:
```
var lookup = new int[][] { };
var q = from i1 in lookup
        from i2 in i1
        orderby i2.ToString().Length
        select new { i1.Length, i2 };
```

The anonymous type a { int[] i1, int i2 } returned from SelectMany is visible in quick info on several occasions:

![2017-11-21 16_16_35-](https://user-images.githubusercontent.com/24472128/33080126-9bef47fc-ced7-11e7-8da3-5530629bb394.png)

![2017-11-21 16_17_15-](https://user-images.githubusercontent.com/24472128/33080112-9372d8dc-ced7-11e7-97e6-b7e851c461c0.png)

sharwell commented:
> orderby clause

Overall, great job here.

:bulb: Consider also binding the commas in a `orderby` clause to the `ThenBy` method(s).

> Explicit range variable types (see cases 7 and 8)

:arrow_right: In the shadowing case, leave `in` referencing the cast. Otherwise the proposed behavior is what users would most likely expect.

#### Summary

Great job with the write-up. I do not anticipate anyone disagreeing with the proposal. Steps to proceed:

1. Copy the proposal to a new issue (as opposed to a pull request), incorporating the update(s) per my specific notes above
2. In your local notes, enumerate the test cases corresponding to each of the cases you've described. Implement as many of the test cases as you reasonably can as part of the pull request. For any remaining ones, itemize them in a comment so we can make an issue to add them later.
3. Let us know when the implementation is ready for review against the spec issue.

MaStr11 commented:
The PR currently implements all features of the proposal #23394 but lacks all tests. A high level code review would be appreciated because the current solution is somewhat scattered across several layers. Problem areas to discuss would be

1. Suppressing Quick-Info for degenerated methods (see also proposal)

This is done [here](https://github.com/dotnet/roslyn/pull/23049/files#diff-b6b3e98a81f47a9ba531457d13efe2aeR305). This location is a little bit arbitrary and the filtering could also be done some layers deeper or higher.

2. Comma is a [bindable token now](https://github.com/dotnet/roslyn/pull/23049/files#diff-1eda1d67f1b7acf2ea04b9af6d0e105aR573)

This is required to support quick info above the comma in a orderby clause.

3. Asymmetric API

This PR introduces the [ISemanticFactsService.GetSymbolInfo(SyntaxNode, SyntaxToken)]( https://github.com/dotnet/roslyn/pull/23049/files#diff-b3318107520cde0627b7ed9621f375b5R258) method. This method is closely related to [SemanticModel.GetSymbolInfo](http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis/Compilation/Extensions.cs,22). Passing the same SyntaxNode to both methods gives sometimes different results depending on the SyntaxToken. This might lead to confusion. I tried to avoid that but wasn't able to do so. See 4. for details:

4. Getting a SymbolInfo for SyntaxNodes that bind to more than one method

Some query clauses SyntaxNodes can bind to more than one method call:
`join int i in l1 on i equals l`
That text is parsed as a `JoinCauseSyntaxNode`. This node binds to a call to `Cast` and to `Join`. The proposal requires to return the `Cast` call for the `in` token and the `Join` call for the other keyword tokens. Other methods hit by this phenomenon: 
  * from i1 in l1 **from int i2 in l2**: FromClauseSyntax binds to SelectMany and Cast and 
  * orderby a1 **,** a2: OrderByClauseSyntax binds to OrderBy and ThenBy.

I went through this solutions to solve that problem and discarded most of them:
1. Use [SymbolInfo.CandidateSymbols](http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis/Compilation/SymbolInfo.cs,32) to return more than one IMethodSymbol
*Discarded:* Would be a abuse of CandidateSymbols as this is for failed bindings.
2. Extend SymbolInfo.CandidateSymbols to support multiple ISymbols.
*Discarded:* Couldn't come up with a good design. Furthermore this is a heavily used struct. 
3. Provide an overload for SemanticModel.GetSymbolInfo(SyntaxToken token, ...)
*Discarded:* Backward compatibility issues.
4. Add a method to SemanticModel: SemanticModel.GetSymbolInfoForToken(SyntaxToken token, ...)
*Discarded:* I tried but it didn't worked because this method would handle the special cases where the token is relevant and than fall-through to the existing SemanticModel.GetSymbolInfo(SyntaxNode node). The fall-through wasn't easy to implement because the node to pass through must be the *bindable* parent. The *bindable* parent is calculated by [`ISyntaxFactsService.GetBindableParent`](http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis.Workspaces/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs,297) which is in a completely different layer. So this would either require to 
  * duplicate the GetBindableParent logic or
  * move the GetBindableParent logic to SemanticModel or
  * Change the signature to SemanticModel.GetSymbolInfoForToken(SyntaxNode node, SyntaxToken token, ...)
Neither of these seemed attractive.
5. Add the SemanticModel.GetSymbolInfoForToken(SyntaxNode node, SyntaxToken token, ...) method to the internal [ISemanticFactsService](https://github.com/dotnet/roslyn/pull/23049/files#diff-76fa299997a1e0ec392f1fe092a56a83R98)

The 5th option is implemented as it seemed the least harmful one. But I'm open for any better solution because passing the SyntaxNode and the SyntaxToken is not what should be done here.

MaStr11 commented:
This PR is ready for review and implements all requirements as defined in the proposal #23394.

sharwell commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -317,6 +317,22 @@ public bool IsBaseConstructorInitializer(SyntaxToken token)
             => node.Kind() == SyntaxKind.QueryExpression;
 
+        public bool IsQueryKeyword(SyntaxToken token)
+            => token.IsKind(
```
💡 This is getting a bit long for an expression-bodied member.

💡 Use a `switch` statement to avoid the array allocation used by `params` argument.

sharwell commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -317,6 +317,22 @@ public bool IsBaseConstructorInitializer(SyntaxToken token)
+                SyntaxKind.EqualsKeyword,
+                SyntaxKind.AscendingKeyword,
+                SyntaxKind.DescendingKeyword) ||
+            (token.IsKind(SyntaxKind.InKeyword) && (token.Parent?.IsKind(SyntaxKind.FromClause, SyntaxKind.JoinClause, SyntaxKind.JoinIntoClause) == true));
```
:bulb: Use `token.IsParentKind` instead of `token.Parent?.IsKind`.

MaStr11 replied:
> Use token.IsParentKind

Couldn't find such a method. There seems to be only a `IsParentKind(this SyntaxNode node` extension method.

sharwell commented on src/Workspaces/Core/Portable/Shared/Extensions/SemanticModelExtensions.cs:
```diff
@@ -201,10 +201,14 @@ private static ISymbol MapSymbol(ISymbol symbol, ITypeSymbol type)
-                                              .WhereAsArray(s => !s.Equals(declaredSymbol))
-                                              .SelectAsArray(s => MapSymbol(s, type));
+
+                var skipSymbolInfoLookup = declaredSymbol != null && declaredSymbol.Kind == SymbolKind.RangeVariable;
```
:bulb: `declaredSymbol.IsKind(SymbolKind.RangeVariable)` can be used since it includes a null check.

sharwell commented on src/Workspaces/Core/Portable/Shared/Extensions/SemanticModelExtensions.cs:
```diff
@@ -201,10 +201,14 @@ private static ISymbol MapSymbol(ISymbol symbol, ITypeSymbol type)
+                allSymbols = skipSymbolInfoLookup
+                    ? ImmutableArray<ISymbol>.Empty
+                    : semanticFacts.GetSymbolInfo(semanticModel, bindableParent, token, cancellationToken)
+                                   .GetBestOrAllSymbols()
```
:bulb: To avoid indenting more than necessary, consider one of the following approaches:

1. Simply don't align on the `.`
2. Move the first `.` to the next line, so all of the `.` characters are aligned but still at a 4-space indent

sharwell commented on src/Workspaces/Core/Portable/Shared/Extensions/SemanticModelExtensions.cs:
```diff
@@ -201,10 +201,14 @@ private static ISymbol MapSymbol(ISymbol symbol, ITypeSymbol type)
+                    : semanticFacts.GetSymbolInfo(semanticModel, bindableParent, token, cancellationToken)
+                                   .GetBestOrAllSymbols()
+                                   .WhereAsArray(s => !s.Equals(declaredSymbol))
+                                   .SelectAsArray(s => MapSymbol(s, type));
```
💭 I need to re-review this from a performance perspective (did not think about it for now)

sharwell commented on src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb:
```diff
@@ -301,6 +301,25 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         End Function
 
+        Public Function IsQueryKeyword(token As SyntaxToken) As Boolean Implements ISyntaxFactsService.IsQueryKeyword
+            Return token.IsKind(
```
:bulb: A `Switch` statement would be more efficient here.

sharwell commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs:
```diff
@@ -296,5 +296,50 @@ public bool IsPartial(ITypeSymbol typeSymbol, CancellationToken cancellationToke
+                    }
+                    break;
+            }
+            if (token.IsKind(SyntaxKind.CommaToken)) //Only in the orderby clause a comma can bind to a symbol.
```
:bulb: Move comment to the line before the `return` statement

sharwell commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs:
```diff
@@ -296,5 +296,50 @@ public bool IsPartial(ITypeSymbol typeSymbol, CancellationToken cancellationToke
+            }
+            if (token.IsKind(SyntaxKind.CommaToken)) //Only in the orderby clause a comma can bind to a symbol.
+            {
+                return new SymbolInfo();
```
💡 Use `default(SymbolInfo)` instead of the default constructor for value types.

sharwell commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -4547,6 +4552,13 @@ private SymbolInfo GetSymbolInfoFromNode(SyntaxNode node, CancellationToken canc
                     return this.GetSymbolInfo(cref, cancellationToken);
                 case SelectOrGroupClauseSyntax selectOrGroupClause:
                     return this.GetSymbolInfo(selectOrGroupClause, cancellationToken);
+                case OrderByClauseSyntax orderByClauseSyntax when orderByClauseSyntax.Orderings[0].AscendingOrDescendingKeyword.Kind() == SyntaxKind.None:
```
❓ Are we sure `orderByClauseSyntax.Orderings[0]` is never null? I'm particularly interested in incomplete code like this:

```csharp
_ = from x in new[] { 0 }
    where x >= 0
    orderby
```

MaStr11 replied:
Test added in commit 648dbae. This seems to be fine.

sharwell commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs:
```diff
@@ -296,5 +296,50 @@ public bool IsPartial(ITypeSymbol typeSymbol, CancellationToken cancellationToke
+
+        public SymbolInfo GetSymbolInfo(SemanticModel semanticModel, SyntaxNode node, SyntaxToken token, CancellationToken cancellationToken)
+        {
+            switch (node)
```
:memo: (note to self) Need to review this block in more detail.

MaStr11 commented:
@sharwell Thanks for the review. Feedback is addressed. Do you have any remarks to my concerns raised in this [comment from 28/11/17](https://github.com/dotnet/roslyn/pull/23049#issuecomment-347583993).

MaStr11 commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -575,7 +607,7 @@ public bool IsTypeNamedDynamic(SyntaxToken token, SyntaxNode parent)
         public bool IsBindableToken(SyntaxToken token)
         {
-            if (this.IsWord(token) || this.IsLiteral(token) || this.IsOperator(token))
+            if (this.IsWord(token) || this.IsLiteral(token) || this.IsOperator(token) || token.Kind() == SyntaxKind.CommaToken)
```
The comma is only bindable as part of an `order by` clause (the 1% case). Maybe this check should reflect this:

```CS
(token.Kind() == SyntaxKind.CommaToken && token.Parent.Kind() == SyntaxKind.OrderByClause)
```

Doing so would make sure, that in the 99% case a comma is not categorized as being bindable and would avoid the reporting of false positives for callers of `IsBindableToken`. Other callers of `IsBindableToken` may also show unintended behavior, if every comma is reported as bindable.

sharwell replied:
👍 I like the idea of restricting this for now.

CyrusNajmabadi replied:
agreed.

MaStr11 commented:
@sharwell What is the status of this PR? It seems to me that this PR is a nice addition to #23970 *QuickInfo should show the captures for lambdas and local functions* as it would allow to show captures in LINQ queries.

sharwell commented:
@Pilchie How do you feel about this one for 15.7?

MaStr11 commented:
> I like the idea of restricting this for now.

I haven't done this yet. The change needed is described [here](https://github.com/dotnet/roslyn/pull/23049#discussion_r171488138).

I'm busy today but will be able to make the change within the next few days.

jinujoseph commented:
@ivanbasov can you review this as well

Pilchie commented:
I would be willing to take this for 15.7 before Preview 3 starts to lock down, provided @MaStr11 can complete the item above, and we have someone pull down the changes locally and buddy test them.

ivanbasov commented on src/Workspaces/Core/Portable/Shared/Extensions/SemanticModelExtensions.cs:
```diff
@@ -230,6 +235,10 @@ private static ISymbol MapSymbol(ISymbol symbol, ITypeSymbol type)
                     }
                 }
             }
+            if (allSymbols.Length == 0 && syntaxFacts.IsQueryKeyword(token))
```
nit: empty line

ivanbasov commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs:
```diff
@@ -296,5 +296,51 @@ public bool IsPartial(ITypeSymbol typeSymbol, CancellationToken cancellationToke
+                    }
+                    break;
+            }
+            //Only in the orderby clause a comma can bind to a symbol.
```
nit: empty line

ivanbasov commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs:
```diff
@@ -296,5 +296,51 @@ public bool IsPartial(ITypeSymbol typeSymbol, CancellationToken cancellationToke
+                case FromClauseSyntax fromClauseSyntax:
+                case JoinClauseSyntax joinClauseSyntax:
+                    var queryClauseSyntax = (QueryClauseSyntax)node;
+                    var qryInfo = semanticModel.GetQueryClauseInfo(queryClauseSyntax, cancellationToken);
```
>qryInfo [](start = 24, length = 7)

'queryInfo' is preferable

CyrusNajmabadi replied:
Yes.  PLease make this change.

ivanbasov commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs:
```diff
@@ -296,5 +296,51 @@ public bool IsPartial(ITypeSymbol typeSymbol, CancellationToken cancellationToke
+                    {
+                        var separators = orderByClauseSyntax.Orderings.GetSeparators().ToImmutableList();
+                        var index = separators.IndexOf(token);
+                        if (index >= 0 && (index + 1) < (orderByClauseSyntax.Orderings.Count))
```
>(orderByClauseSyntax.Orderings.Count) [](start = 56, length = 37)

why do we need parentheses here?

ivanbasov commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs:
```diff
@@ -296,5 +296,51 @@ public bool IsPartial(ITypeSymbol typeSymbol, CancellationToken cancellationToke
+                            }
+                        }
+                    }
+                    break;
```
empty line

CyrusNajmabadi commented:
Just to check, did this get compiler signoff?  There are changes to the compiler code.  Thanks!

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs:
```diff
@@ -296,5 +296,51 @@ public bool IsPartial(ITypeSymbol typeSymbol, CancellationToken cancellationToke
+                    if (hasCastInfo && hasOperationInfo)
+                    {
+                        return token.IsKind(SyntaxKind.InKeyword) ? qryInfo.CastInfo : qryInfo.OperationInfo;
+                    }
```
the logic here could use good commenting as to why you're doing what your'e doing.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -317,6 +317,38 @@ public bool IsBaseConstructorInitializer(SyntaxToken token)
+                case SyntaxKind.DescendingKeyword:
+                    return true;
+                case SyntaxKind.InKeyword:
+                    switch (token.Parent?.Kind())
```
tokens always have parents in our tree.  so ?  isn't necessary.

MaStr11 commented:
PR feedback is addressed. I will need some help with fixing the merge conflicts and changing the target branch (I failed badly the last few times I tried this myself).

sharwell commented:
@MaStr11 No problem I'm on it!

MaStr11 commented:
Thanks @sharwell for the rebasing. Would you also take care of this 

> Just to check, did this get compiler signoff? There are changes to the compiler code. 

@CyrusNajmabadi is right. The changes I made there were copy and paste. Someone should verify it.

jinujoseph commented:
request review from  @dotnet/roslyn-compiler

jinujoseph commented:
@richaverma1 , can we get this tested before we get this in ?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -491,6 +491,11 @@ internal override IOperation CloneOperationCore(IOperation operation)
+        /// Gets the semantic information associated with a LINQ query syntax clause.
+        /// </summary>
+        public abstract SymbolInfo GetSymbolInfo(QueryClauseSyntax node, CancellationToken cancellationToken = default(CancellationToken));
+
```
It is not clear why this is needed, we have existing API that return information about QueryClauseSyntax, for example GetQueryClauseInfo.

AlekseyTs commented:
Is existing GetQueryClauseInfo API not sufficient? How exactly is it not sufficient. I would prefer to see no changes on the compiler for this feature. If that is not possible, we should understand why. Then at the very least we should have targeted semantic model tests for all new code paths for all possible query clauses, etc.

MaStr11 commented:
@AlekseyTs 

> Is existing GetQueryClauseInfo API not sufficient?

It is and I reverted the changes in the compiler layer (I raised my concerns (*the current solution is somewhat scattered across several layers*) back in [November](https://github.com/dotnet/roslyn/pull/23049#issuecomment-347583993) but it seemed to be no problem until now.

Nevertheless the method [CSharpSemanticModel .GetSymbolInfoFromNode](
https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs#L4534-L4554) is still quite odd. It handles `SelectOrGroupClauseSyntax` and `OrderingSyntax` but nothing else related to queries. I still need the logic I placed there but I moved it up to `ISemanticFactsService` in 87f9f6f.

Because of that, another concern I raised in the [same comment](https://github.com/dotnet/roslyn/pull/23049#issuecomment-347583993) becomes even more relevant:

> 3. Asymmetric API
>
> This PR introduces the ISemanticFactsService.GetSymbolInfo(SyntaxNode, SyntaxToken) method. This method is closely related to SemanticModel.GetSymbolInfo. Passing the same SyntaxNode to both methods gives sometimes different results depending on the SyntaxToken.

Now both method give also different results depending on the SyntaxNode (e.g. WhereClauseNode). I still think that this imbalance should be addressed.

AlekseyTs commented:
> Nevertheless the method CSharpSemanticModel .GetSymbolInfoFromNode is still quite odd. It handles SelectOrGroupClauseSyntax and OrderingSyntax but nothing else related to queries.

This method is an internal helper, as long as public API behaves as expected, it does what it is supposed to do.
I cannot comment on issues with ISemanticFactsService, that interface looks like an implementation detail in IDE layer and I let IDE team to decide whether there is an imbalance and what changes should be made to that interface to address it, if any.
As long as you don't have any changes in compilers, there is no requirement to get a sign-off from compiler team.

CyrusNajmabadi commented:
> Now both methods give also different results depending on the SyntaxNode (e.g. WhereClauseNode). I still think that this imbalance should be addressed.

I'm ok with the imbalance.  SemanticModel is clearly defined as to how it behaves, with expectations on the caller to call the right methods to get the appropriate information back.   ISemanticFactsService.GetSymbolInfo is intended to be a bit of a muddier API.  It's essentially the "if the user were to ask what this location means, what would i give back?"  We have it so that every feature doesn't have to reproduce that logic, and so that things like quick-info/goto-def/etc. can be consistent in their view of the world.  i.e. we can change one locaiton and have several features get updated appropriately.

MaStr11 commented:
> I'm ok with the imbalance.

Good. This imbalance is easy to spot while you are coding but nearly impossible to recognize during a review. I just didn't wanted this to slip through unnoticed.

Besides that I still don't get the reasoning behind [CSharpSemanticModel .GetSymbolInfoFromNod](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs#L4534-L4554). It handles some arbitrary SyntaxNodes and ignores others (in respect to query syntax).

CyrusNajmabadi commented:
>  CSharpSemanticModel .GetSymbolInfoFromNod

Internal implementation detail.  It's possible it could benefit from being refactored.  But that's totally orthogonal to any convos here.  If you want to, feel free to create a new PR that addresses that.  There you can clean things up, and make your case for why it's better than what the compiler has today.  If the compiler team agrees, then it can be merged in :)

AlekseyTs commented:
@dotnet/roslyn-ide Please squash commits before merging PRs, let's keep history nice, clean and meaningful. Thanks

jcouv commented:
@MaStr11 Showing captures in queries sounds like a good idea. I'll let @sharwell and @dotnet/roslyn-ide confirm this question from user experience perspective.

I assume that the compiler's region analysis works as expected on the query lambdas. Note that queries involve a special kind of local ("ranged variables") so surprises could be possible.

My recommendation would be to do a separate PR after this one.

jcouv commented on src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/Providers/AbstractSemanticQuickInfoProvider.cs:
```diff
@@ -280,11 +280,10 @@ internal abstract partial class AbstractSemanticQuickInfoProvider : AbstractQuic
             SyntaxToken token,
             CancellationToken cancellationToken)
         {
+            var syntaxFacts = document.Project.LanguageServices.GetService<ISyntaxFactsService>();
```
It seems `syntaxFacts` is only used at line 319 or below. Consider declaring around there, since there is possibility of returning early (without needing `syntaxFacts`).

MaStr11 replied:
It turned out there was another call to `document.GetLanguageService<ISyntaxFactsService>()` just a few lines down. I changed it and left the `var syntactsFacts=` as is because there isn't any `return` between the initialization and the first use anymore.

jcouv commented on src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs:
```diff
@@ -3015,6 +3015,506 @@ void M()
+        {
+            await TestInMethodAsync(
+@"
+            var q = $$from double i in new int[0]
```
Could you add a similar test without the `double` cast, ie. for the vanilla `$$from` scenario?

MaStr11 replied:
Sure I can. But there is nothing to be shown in that case.

MaStr11 replied:
Done db7afce.

MaStr11 commented:
Thanks @jcouv for having a look.

> Showing captures in queries sounds like a good idea. ... My recommendation would be to do a separate PR after this one.

I have looked into it and while it seems possible (could it get working on a where clause) it is not as straight forward as I was hoping (this eventually needs some planning as it isn't obvious where to show the capture info). So this should definitely be done in another PR.

jcouv commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -344,6 +344,8 @@ public bool IsQueryKeyword(SyntaxToken token)
                             return false;
                     }
+                case SyntaxKind.WhereKeyword:
+                    return token.Parent.Kind() == SyntaxKind.WhereClause; // false for e.g. type parameter constraint
```
Iteration 11 is adding some logic, but I don't see any impacted tests. Are you planning to add such tests? Thanks

MaStr11 replied:
AFIK there aren't any test for `CSharpSyntaxFactsService`. This service is tested indirectly by it's usage in other services (e.g. QuickInfo). @CyrusNajmabadi knows more about it.

CyrusNajmabadi replied:
Yes.  We do not test this service directly.  We're ok with it being indirectly covered through tests of other features.  We want to ensure that the experience users have of features is what is tested.  We do not believe in testing the underlying implementation-detail pieces as that just creates brittle tests for us where we have to go update tests when impl-details change, even if the user experience stays the same.

sharwell commented:
@Pilchie @MattGertz for approval

sharwell commented:
@MaStr11 Please let me know if you want me to squash the commits for this pull request during the merge. If I get approval to merge and don't hear back from you, I'm going to merge the it as-is (using a merge commit and not squashing commits).

MaStr11 commented:
Squashing was requested by the compiler team so my back and forth in their layer isn't part of the history. I'm fine with squashing but I don't know how to do it. It would be nice if you could do this for me.

sharwell commented:
@MaStr11 Thanks, by your request I'll apply the squash during the merge

MattGertz commented:
Very cool.  Scenario meets the ASK bar.
